MDV2 datasets

emtwo opened the pull request:
This PR has a few things:

* A script for generating the clients_daily_scalar_aggregates_v1 dataset
* The actual sql produced by the query
* Basic test inputs/outputs
* client_probe_counts.sql query

Notes:
* The plan is to query main_ping which will have several tables (e.g. one for histograms, scalars, booleans, etc). Each main_ping table would have an analogous mdv2 table that aggregates it by client
* clients_daily_scalar_aggregates_v1 looks very similar to clients_daily_v7, I got some of the fundamentals about document_id deduping etc. from there.
* client_probe_counts.sql is still rough work so I'm mostly hoping for a review pass on the other work aside from this
* The test isn't very elaborate, didn't want to put too much into it until this gets through first pass of review
* The script to generate clients_daily_scalar_aggregates_v1.sql finds the intersection between relevant probes found in the probe info service and those that exist in main summary. This might not be the best way to do it (i.e. should we use an api call instead of `bq show` to find the schema and know which are scalars?) Hoping to get feedback on this.

robhudson commented on sql/telemetry/clients_aggregates_v1.sql:
```diff
@@ -0,0 +1,166 @@
+                aggregate.metric_type,
+                aggregate.key,
+                aggregate.agg_type)
+  ),
```
This looks like the identical query as the `scalars_and_booleans` one above it. I'm curious why it's separated out?

robhudson replied:
Aha! It's pulling from a different table. I scanned over the `FROM` line.

robhudson commented on templates/telemetry/clients_daily_histogram_aggregates_v1.sql.py:
```diff
@@ -0,0 +1,444 @@
+          Object.keys(node).map(function(key) {
+            value = node[key].toString();
+            z.push(key + ":" + value);
+          });
```
Since you want both the key and value, this could be written as:
```
for (let [key, value] of Object.entries(node)) {
  z.push(`${key}:${value}`);
}
```

I couldn't find any docs on what JS spec these UDFs support so maybe they don't like the templated strings (from ES2015)?

robhudson commented on templates/telemetry/clients_daily_histogram_aggregates_v1.sql.py:
```diff
@@ -0,0 +1,444 @@
+                UNNEST(histograms) AS histogram,
+                UNNEST(get_keyval_pairs(JSON_EXTRACT(histogram, "$.values"))) AS keyval
+              WHERE histogram IS NOT NULL
+              GROUP BY key));
```
This is cool :tada:

robhudson commented on templates/telemetry/clients_daily_histogram_aggregates_v1.sql.py:
```diff
@@ -0,0 +1,444 @@
+            DATE(submission_timestamp) as submission_date,
+            client_id,
+            normalized_os as os,
+            SPLIT(application.version, '.')[OFFSET(0)] AS app_version,
```
Nice, I see where this is happening now. :+1:

robhudson commented on templates/telemetry/clients_daily_histogram_aggregates_v1.sql.py:
```diff
@@ -0,0 +1,444 @@
+        return
+
+    for histogram in histograms_field.get("fields", []):
+        main_summary_histograms.add(histogram["name"])
```
Unless I'm reading this wrong, you're providing a list as the default value to the `.get()`, but in the loop you're treating `histogram` as a dict, which would result in a `TypeError` if the key `fields` didn't exist.

emtwo replied:
good catch!

robhudson commented on templates/telemetry/clients_daily_histogram_aggregates_v1.sql.py:
```diff
@@ -0,0 +1,444 @@
+                for x in data.keys()
+                if x.startswith("histogram/")
+            ]
+        )
```
You can use a set comprehension instead of creating a list and calling `set()` on it...
```
histogram_probes = {
    x.replace("histogram/", "").replace(".", "_").lower()
    for x in data.keys()
    if x.startswith("histogram/")
}
```

emtwo commented:
This PR is fairly hefty, so I'm going to give a high-level overview of what's included here. Here's the data set proposal link for reference as it's mostly what was implemented: https://docs.google.com/document/d/1rX8HBnpc0hH-hIG_v21f-aPBfFFc5LF4pzD3v3WRW5M/edit

At a high level, there is a 3-step process to go from main ping -> `client_probe_counts_v1` which contains data that is directly plot-able by a front-end.
1. `clients_daily_aggregates` is just what it sounds like - aggregation by day and by client over a set of dimensions.
2. `clients_aggregates` is like `clients_daily_aggregates` but is not aggregated by date.
3. `client_probe_counts` is a set of counts of clients per probe as a plot-able distribution

So far, in this PR, these steps are here only for sum/avg/min/max for scalars and for normalized histograms. Percentiles/median will need to be added for the evolution view (https://github.com/mozilla/bigquery-etl/issues/322)

#### Clients Daily Aggregates
There are two python scripts for generating this dataset, `clients_daily_scalar_aggregates_v1.sql.py` and `clients_daily_histogram_aggregates_v1.sql.py` which are used to generate the scalar and histogram aggregations respectively.

A command line param, `--agg-type` can be set to one of `{scalar`, `keyed-scalar`, `keyed-boolean`} in the scalar script and one of {`histograms`, `keyed_histograms`} in the histogram script. The sql to generate each of these aggregation types is slightly different but the output of all of all scalar aggregations will have the same format and the output of all histogram aggregations will have the same format.

#### Clients Aggregates
This is an intermediate dataset which only exists to break up computations into smaller pieces to reduce computation time of each. In theory, all computation done for clients aggregates can be done in `client_probe_counts_v1.sql`.

In fact, `clients_aggregates` is an intermediate table in `client_probe_counts_v1.sql` that is only for histograms. `clients_aggregates_v1.sql ` is actually only for scalar (keyed/booleans/regular) computations. Handling booleans needed a bit of extra work so I put it in this file.

I realize this split is probably confusing so I'm open to re-evaluating putting all clients aggregates computations in `client_probe_count_v1.sql`, or any other ideas.

#### Client Probe Counts
This dataset brings it all together with both all histogram and all scalar aggregations in one place. It isn't quite complete yet because for scalars there are a variety of combinations of dimensions set to `NULL` in the subqueries (to include all values for that dimension). This isn't there yet for histograms - I've only added one subquery so far for histograms.

emtwo commented on sql/telemetry/clients_aggregates_v1.sql:
```diff
@@ -0,0 +1,166 @@
+
+    SELECT *
+    FROM keyed_scalars
+    WHERE metric_type = "keyed-scalar-boolean"),
```
OH! You actually made me realize a mistake... 
There should only be one table called `clients_daily_scalar_aggregates_v1` which contains keyed scalars as well. So this would naturally be less redundant.

Thank you!!

fbertsch commented:
A lot of the comments I made are improvements and extensions on what you have. Blocking this change for an MVP of just main process histograms/scalars is:

- Remove histogam re-bucketing, and only have the normalized-sum
- Fix the bucketing algos
- Add `GROUPING SETS` for histograms (using the `UNION ALL` syntax)
- Add where clause limits for recent versions

See [the relevant section](https://docs.google.com/document/d/1zgjdaRSL-0cDctxMEjLBmCwSw6F0vDfRttNh2B_qvTw/edit#heading=h.bmodko434hzh) for how we proposed ignoring old data.

emtwo commented:
@fbertsch I've updated the PR with the 4 critical changes (and some others). I've filed issues for changes I haven't yet made.

I've changed the structure a bit to make these changes work. The where clause limits + grouping sets caused an out of memory error for the original `client_probe_counts`. 

So I've reorganized it such that there is one query for histogram client probe counts and one for scalar. Each one include the relevant "clients_aggregates" computations. All client_probe_counts results will still end up in one table.

fbertsch commented on sql/telemetry/client_probe_counts_v1/query.sql:
```diff
@@ -177,7 +177,23 @@ RETURNS ARRAY<STRUCT<key STRING, value FLOAT64>> AS (
+    WHERE DATE(submission_timestamp) > DATE_SUB(CURRENT_DATE(), INTERVAL 90 DAY)
+    AND normalized_channel IN ("nightly", "beta", "release")
+    GROUP BY 1, 2
+    HAVING COUNT(*) > 1000
```
I'm casting a wary eye on this `100` and the `limit 100`. I believe we should increase the value here to say, `1000`, and remove the `limit` altogether.

emtwo replied:
oops! missed removing that limit. nice catch! Thanks! The `having count(*) > 1000` is in the latest commit

fbertsch commented on sql/telemetry/clients_aggregates_v1/query.sql:
```diff
@@ -10,12 +34,23 @@ WITH scalar_aggregates AS (
+    clients_daily_scalar_aggregates_v1 AS scalar_aggs
   CROSS JOIN
     UNNEST(scalar_aggregates)
+  LEFT JOIN latest_versions
```
Are `scalar_aggregates` being limited by version here? I see a join but no where clause.

emtwo replied:
lol indeed. I missed this and just pushed it now. thanks!

fbertsch commented on sql/telemetry/client_histogram_probe_counts_v1/query.sql:
```diff
@@ -0,0 +1,577 @@
+  LEFT JOIN latest_versions
+  ON latest_versions.channel = hist_aggs.channel
+  WHERE ARRAY_LENGTH(value) > 0
+  AND ((hist_aggs.channel = 'release' AND CAST(app_version AS INT64) >= (latest_version - 2))
```
Hmm, it seems to me we should also be limiting by `date` since that is how `clients_daily_*` is partitioned. We limit to 90 days for `latest_versions`, we could do that, or we could find the min date for the channel/versions.

fbertsch commented:
@emtwo I reviewed per-commit,  but I believe each comment is still relevant for the final commit. If not feel free to ignore!

emtwo commented on sql/telemetry/client_probe_counts_v1/query.sql:
```diff
@@ -81,8 +81,17 @@ RETURNS ARRAY<STRUCT<key STRING, value FLOAT64>> AS (
 );
 
-CREATE TEMP FUNCTION udf_get_buckets(min INT64, max INT64, num INT64, metric_type STRING)
+CREATE TEMP FUNCTION udf_to_string_arr(buckets ARRAY<INT64>)
```
So right now the "histogram" for a boolean uses the keys `['always','never','sometimes']` which are strings. In order to keep both boolean and scalar (and anything else) in the same table, I cast all keys to strings.

Any thoughts on an alternative? I guess we could make always/sometimes/never be represented by 0/1/2. Just figured it would be nice for the UI (or any user querying the data) to have these to access directly

fbertsch replied:
okay, I see. It is a bit confusing right now but we can think through an alternative at a later stage.
